drivers/usb/device/usb_dc_numaker.c

```diff
@@ -0,0 +1,1931 @@
+		}
+
+		if (read_bytes) {
+			*read_bytes = data_len_act;
+		}
+	} else if (max_data_len) {
+		LOG_ERR("Null data but non-zero data length");
+		rc = -EINVAL;
+		goto cleanup;
+	} else {
+		if (read_bytes) {
+			*read_bytes = numaker_usbd_ep_fifo_used(ep_cur);
+		}
+	}
+
+	/* Suppress further USB_DC_EP_DATA_OUT events by replying NAK or disabling interrupt
+	 *
+	 * For USBD, further control is unnecessary because NAK is automatically replied until
+	 * next USBD_SET_PAYLOAD_LEN().
+	 */
+
+cleanup:
+
+	numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_read_continue(uint8_t ep_addr)
+{
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	int rc = 0;
+
+	numaker_usbd_lock(dev);
+
+	/* Bind EP context to EP address */
+	struct numaker_usbd_ep *ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_addr);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep_addr=0x%02x", ep_addr);
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	if (!USB_EP_DIR_IS_OUT(ep_addr)) {
+		LOG_ERR("Invalid EP address 0x%02x for read", ep_addr);
+		rc = -EINVAL;
+		goto cleanup;
+	}
+
+	/* Avoid duplicate NAK clear */
+	if (ep_cur->nak_clr) {
+		rc = 0;
+		goto cleanup;
+	}
+
+	/* For one-shot implementation, don't trigger next DATA OUT, or overwrite. */
+	if (numaker_usbd_ep_fifo_used(ep_cur)) {
+		goto cleanup;
+	}
+
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	numaker_usbd_ep_trigger(ep_cur, ep_cur->ep_mps);
+
+cleanup:
+
+	numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+int usb_dc_ep_mps(const uint8_t ep_addr)
+{
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	int rc = 0;
+	uint16_t ep_mps = 0;
+
+	numaker_usbd_lock(dev);
+
+	/* Bind EP context to EP address */
+	struct numaker_usbd_ep *ep_cur = numaker_usbd_ep_mgmt_bind_ep(dev, ep_addr);
+
+	if (!ep_cur) {
+		LOG_ERR("Bind EP context: ep_addr=0x%02x", ep_addr);
+		rc = -ENOMEM;
+		goto cleanup;
+	}
+
+	__ASSERT_NO_MSG(ep_cur->ep_mps_valid);
+	ep_mps = ep_cur->ep_mps;
+
+cleanup:
+
+	numaker_usbd_unlock(dev);
+
+	return rc == 0 ? ep_mps : rc;
+}
+
+int usb_dc_wakeup_request(void)
+{
+	LOG_DBG("Remote wakeup");
+
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	int rc = 0;
+
+	numaker_usbd_lock(dev);
+
+	numaker_usbd_remote_wakeup(dev);
+
+	numaker_usbd_unlock(dev);
+
+	return rc;
+}
+
+static int usb_dc_numaker_init(void)
+{
+	struct usb_dc_numaker_device *dev = usb_dc_numaker_device_get();
+	struct usb_dc_numaker_data *data = dev->data;
+	int rc = 0;
+
+	/* Initialize all fields to zero */
+	memset(data, 0x00, sizeof(*data));
+
+	k_mutex_init(&data->sync_mutex);
+
+	/* Set up interrupt top/bottom halves processing */
+
+	k_msgq_init(&data->msgq, (char *)data->msgq_buf, sizeof(struct numaker_usbd_msg),
+		    CONFIG_USB_DC_NUMAKER_MSG_QUEUE_SIZE);
+
+	k_thread_create(&data->msg_hdlr_thread, data->msg_hdlr_thread_stack,
+			CONFIG_USB_DC_NUMAKER_MSG_HANDLER_THREAD_STACK_SIZE,
+			numaker_usbd_msg_hdlr_thread_main, dev, NULL, NULL, K_PRIO_COOP(2), 0,
+			K_NO_WAIT);
+
```

Comment:
I'd recommend a `k_thread_name_set(&data->msg_hdlr_thread, dev->name);` here (or a fixed name, whatever  you want just give it a name)

Reply:
added